File: dp/cpp/lazy_initialization.cpp
// Lazy Initialization — Concurrency / Creational Pattern
//
// Intent: defer the creation of an expensive resource until it is first used,
//         and ensure it is only created once even under concurrent access.
//
// When to use:
//   - The resource is costly to build and may never be needed
//   - Thread safety is required (singleton, shared caches, connection pools)
//
// When NOT to use:
//   - The resource is cheap — just construct it eagerly
//   - Startup latency hiding is acceptable — eager init keeps runtime predictable

#include <iostream>
#include <mutex>
#include <thread>
#include <vector>
#include <memory>
#include <optional>
#include <chrono>

class Database {
public:
    Database() {
        std::cout << "  [Database] opening connection (expensive!)...\n";
        std::this_thread::sleep_for(std::chrono::milliseconds(50));
    }
    void query(const std::string& q) const {
        std::cout << "  [Database] " << q << "\n";
    }
};

// ============================================================================
// Example 1: std::call_once — the canonical modern approach
// ============================================================================
class LazyCallOnce {
public:
    Database& get() {
        std::call_once(flag_, [this] { db_ = std::make_unique<Database>(); });
        return *db_;
    }
private:
    std::once_flag              flag_;
    std::unique_ptr<Database>   db_;
};

// ============================================================================
// Example 2: function-local static — simplest thread-safe lazy
// ============================================================================
Database& get_global_db() {
    static Database db;   // thread-safe construction guaranteed by C++11
    return db;
}

// ============================================================================
// Example 3: std::optional for member-level lazy without pointer overhead
// ============================================================================
class Service {
public:
    void do_work() {
        if (!cache_) {
            cache_.emplace();    // default-construct on first use
            cache_->push_back("initialized");
        }
        cache_->push_back("used");
        std::cout << "  cache size: " << cache_->size() << "\n";
    }
private:
    std::optional<std::vector<std::string>> cache_;
};

int main() {
    std::cout << "=== Lazy Initialization Pattern ===\n";

    std::cout << "\n--- call_once ---\n";
    LazyCallOnce lazy;
    std::vector<std::thread> ts;
    for (int i = 0; i < 4; ++i) {
        ts.emplace_back([&lazy, i] {
            lazy.get().query("query from thread " + std::to_string(i));
        });
    }
    for (auto& t : ts) t.join();

    std::cout << "\n--- function-local static ---\n";
    get_global_db().query("select 1");

    std::cout << "\n--- optional ---\n";
    Service svc;
    svc.do_work();
    svc.do_work();

    std::cout << "\nKey points:\n";
    std::cout << " * call_once: explicit, works for member-level lazy init\n";
    std::cout << " * static local: simplest for global/singleton resources\n";
    std::cout << " * optional: avoids heap allocation for deferred member objects\n";
    return 0;
}

File: dp/rust/src/lazy_initialization.rs
// Lazy Initialization Pattern in Rust
// Defer expensive computation until first access

use std::cell::OnceCell;
use std::sync::OnceLock;

// Thread-local lazy (single-threaded)
struct LazyResource {
    data: OnceCell<Vec<u8>>,
}

impl LazyResource {
    fn new() -> Self {
        println!("[LazyResource] Created (data NOT loaded)");
        LazyResource {
            data: OnceCell::new(),
        }
    }

    fn get_data(&self) -> &[u8] {
        self.data.get_or_init(|| {
            println!("[LazyResource] Loading data (expensive)...");
            vec![1, 2, 3, 4, 5]
        })
    }
}

// Thread-safe lazy (static)
fn global_config() -> &'static String {
    static CONFIG: OnceLock<String> = OnceLock::new();
    CONFIG.get_or_init(|| {
        println!("[Global] Loading config...");
        "production".to_string()
    })
}

// Lazy with Option + mutable access
struct LazyComputed {
    input: i32,
    cached: Option<i64>,
}

impl LazyComputed {
    fn new(input: i32) -> Self {
        LazyComputed { input, cached: None }
    }

    fn compute(&mut self) -> i64 {
        if let Some(val) = self.cached {
            return val;
        }
        println!("[LazyComputed] Computing for input {}...", self.input);
        let result = (self.input as i64) * (self.input as i64);
        self.cached = Some(result);
        result
    }
}

fn main() {
    // OnceCell lazy
    println!("=== OnceCell ===");
    let resource = LazyResource::new();
    println!("Before access...");
    println!("Data: {:?}", resource.get_data());
    println!("Data again: {:?}", resource.get_data()); // Not loaded again

    // OnceLock global lazy
    println!("\n=== OnceLock (global) ===");
    println!("Config: {}", global_config());
    println!("Config again: {}", global_config());

    // Mutable lazy
    println!("\n=== Option cache ===");
    let mut calc = LazyComputed::new(42);
    println!("Result: {}", calc.compute());
    println!("Result again: {}", calc.compute());
}
